hardware/kubernetes_client.go

```diff
@@ -0,0 +1,255 @@
+		return KubernetesClientConfig{}, err
+	}
+
+	return KubernetesClientConfig{
+		Config:    config,
+		Namespace: namespace,
+	}, nil
+}
+
+// ExportableK8sHardwareHardwareFromK8sObject creates an ExportableK8sHardware from tinkHardware.
+func ExportableK8sHardwareHardwareFromK8sObject(tinkHardware *tinkv1alpha1.Hardware) *ExportableK8sHardware {
+	hw := &ExportableK8sHardware{
+		Hardware: tinkHardware,
+		Metadata: ExportableK8sHardwareMetadata{
+			Userdata: tinkHardware.Spec.UserData,
+			Instance: ExportableK8sHardwareMetadataInstance{
+				ID:        tinkHardware.Spec.Metadata.Instance.ID,
+				Hostname:  tinkHardware.Spec.Metadata.Instance.Hostname,
+				Plan:      tinkHardware.Spec.Metadata.Facility.PlanSlug,
+				Factility: tinkHardware.Spec.Metadata.Facility.FacilityCode,
+				Tags:      tinkHardware.Spec.Metadata.Instance.Tags,
+				SSHKeys:   tinkHardware.Spec.Metadata.Instance.SSHKeys,
+				OperatingSystem: ExportableK8sHardwareMetadataInstanceOperatingSystem{
+					Slug:     tinkHardware.Spec.Metadata.Instance.OperatingSystem.Slug,
+					Distro:   tinkHardware.Spec.Metadata.Instance.OperatingSystem.Distro,
+					Version:  tinkHardware.Spec.Metadata.Instance.OperatingSystem.Version,
+					ImageTag: tinkHardware.Spec.Metadata.Instance.OperatingSystem.ImageTag,
+				},
+			},
+		},
+	}
+
+	for _, ip := range tinkHardware.Spec.Metadata.Instance.Ips {
+		hw.Metadata.Instance.Network.Addresses = append(
+			hw.Metadata.Instance.Network.Addresses,
+			ExportableK8sHardwareMetadataInstanceNetworkAddress{
+				Address:       ip.Address,
+				AddressFamily: ip.Family,
+				Public:        ip.Public,
+			},
+		)
+	}
+
+	return hw
+}
+
+// ExportabelK8sHardware satisfies Hardware such that hardware metadata can be exported and queried.
+type ExportableK8sHardware struct {
+	Hardware *tinkv1alpha1.Hardware        `json:"-"`
+	Metadata ExportableK8sHardwareMetadata `json:",omitempty"`
+}
+
+type ExportableK8sHardwareMetadata struct {
+	Userdata *string                               `json:",omitempty"`
+	Instance ExportableK8sHardwareMetadataInstance `json:",omitempty"`
+}
+
+type ExportableK8sHardwareMetadataInstance struct {
+	ID              string                                               `json:",omitempty"`
+	Hostname        string                                               `json:",omitempty"`
+	Plan            string                                               `json:",omitempty"`
+	Factility       string                                               `json:",omitempty"`
+	Tags            []string                                             `json:",omitempty"`
+	SSHKeys         []string                                             `json:"ssh_keys,omitempty"`
+	OperatingSystem ExportableK8sHardwareMetadataInstanceOperatingSystem `json:"operating_system,omitempty"`
+	Network         ExportableK8sHardwareMetadataInstanceNetwork         `json:",omitempty"`
+}
+
+type ExportableK8sHardwareMetadataInstanceOperatingSystem struct {
```

Comment:
Seems we could be using the existing types for most of this to prevent introducing diff and drift:

https://github.com/tinkerbell/tink/blob/main/pkg/apis/core/v1alpha1/hardware_types.go#L156-L162

Reply:
The existing types have a different structure so while they seem similar, they don't satisfy the Hegel metadata expectations from the `Export()` call. 

Moreover, I think its beneficial to have a clear distinction between Hegel metadata structure and back-end storage format as it allows us to change the structure of the back-end data without impacting hegel given there's an explicit translation.

Reply:
I think this is OK. Please unresolve if needed.